data-prepper-plugins/aws-lambda/src/main/java/org/opensearch/dataprepper/plugins/lambda/processor/LambdaProcessor.java

```diff
@@ -137,41 +139,44 @@ public Collection<Record<Event>> doExecute(Collection<Record<Event>> records) {
             // If the condition is false, add the event to resultRecords as-is
-            if (whenCondition != null && !expressionEvaluator.evaluateConditional(whenCondition, event)) {
+            if (whenCondition != null && !expressionEvaluator.evaluateConditional(whenCondition,
+                event)) {
                 resultRecords.add(record);
                 continue;
             }
             recordsToLambda.add(record);
         }
-        try {
-            resultRecords.addAll(
-                    lambdaCommonHandler.sendRecords(recordsToLambda, lambdaProcessorConfig, lambdaAsyncClient,
-                            new OutputCodecContext(),
-                            (inputBuffer, response) -> {
-                                Duration latency = inputBuffer.stopLatencyWatch();
-                                lambdaLatencyMetric.record(latency.toMillis(), TimeUnit.MILLISECONDS);
-                                numberOfRecordsSuccessCounter.increment(inputBuffer.getEventCount());
-                                numberOfRequestsSuccessCounter.increment();
-                                return convertLambdaResponseToEvent(inputBuffer, response);
-                            },
-                            (inputBuffer) -> {
-                                Duration latency = inputBuffer.stopLatencyWatch();
-                                lambdaLatencyMetric.record(latency.toMillis(), TimeUnit.MILLISECONDS);
-                                numberOfRecordsFailedCounter.increment(inputBuffer.getEventCount());
-                                numberOfRequestsFailedCounter.increment();
-                                return addFailureTags(inputBuffer.getRecords());
-                            })
-
-            );
-        } catch (Exception e) {
-            LOG.info("Exception in doExecute");
-            numberOfRecordsFailedCounter.increment(recordsToLambda.size());
-            resultRecords.addAll(addFailureTags(recordsToLambda));
+
+        Map<Buffer, CompletableFuture<InvokeResponse>> bufferToFutureMap = LambdaCommonHandler.sendRecords(
+            recordsToLambda, lambdaProcessorConfig, lambdaAsyncClient,
+            new OutputCodecContext());
+
+        for (Map.Entry<Buffer, CompletableFuture<InvokeResponse>> entry : bufferToFutureMap.entrySet()) {
+            CompletableFuture<InvokeResponse> future = entry.getValue();
+            Buffer inputBuffer = entry.getKey();
+            try {
+                InvokeResponse response = future.join();
+                Duration latency = inputBuffer.stopLatencyWatch();
+                lambdaLatencyMetric.record(latency.toMillis(), TimeUnit.MILLISECONDS);
+                if (isSuccess(response)) {
+                    numberOfRecordsSuccessCounter.increment(inputBuffer.getEventCount());
+                    numberOfRequestsSuccessCounter.increment();
+                    resultRecords.addAll(convertLambdaResponseToEvent(inputBuffer, response));
```

Comment:
I was going to do this in my PR but please move this line up by two lines so that if convertLambdaReponseToEvent fails, the counters do not get incremented.

Reply:
Discussed with Krishan, he will address this change in his next PR